packages/uma-auth-client/src/hooks/useOAuth.tsx

```diff
@@ -0,0 +1,100 @@
+import {auth, nwc, webln} from '@getalby/sdk';
+import { useEffect, useState } from "react";
+import { getUmaDomain } from "src/utils/getUmaDomain";
+import { useUser } from "./useUser";
+import { create } from 'zustand';
+import { persist } from "zustand/middleware";
+import { getNwcUmaDomain } from 'src/utils/getNwcUmaDomain';
+
+type UmaAuthToken = {
+  access_token: string,
+  refresh_token: string,
+  expires_at: number,
+  nwc_connection_uri: string;
+  commands: string[];
+  budget: string;
+  nwc_expires_at: string;
+};
+
+interface TokenState {
+  accessToken: string;
+  refreshToken: string;
+  expiresAt: number;
+}
+
+interface OAuthState {
+  codeVerifier?: string;
+  token?: TokenState;
+  identityNpub?: string;
+  identityRelayUrl?: string;
+  redirectUri?: string;
+  nwcConnectionUri?: string;
+  setToken: (token: TokenState) => void;
+  performOAuth: (uma: string) => Promise<void>;
+}
+
+export const useOAuth = create<OAuthState>()(
+  persist(
+    (set, get) => ({
+      setToken: (token) => set({
+        token,
+      }),
+      performOAuth: async (uma) => {
+        const state = get();
+        const result = await performOAuth(state, uma);
+        set(result);
+      },
+    }),
+    {
+      name: "uma-connect",
+      partialize: (state) => ({
+        nwcConnectionUri: state.nwcConnectionUri,
+      })
+    }
+  )
+);
+
+const performOAuth = async (state: OAuthState, uma: string) => {
+  const { identityNpub, identityRelayUrl, redirectUri, codeVerifier, token, nwcConnectionUri } = state;
+
+  // If we have a connection URI and a token that hasn't expired, we don't need to do anything
+  if (nwcConnectionUri && token && (!token.expiresAt || token.expiresAt < Date.now())) {
+    return state;
+  }
+
+  const authClient = new auth.OAuth2User({
+    client_id: `${identityNpub} ${identityRelayUrl}`,
+    callback: redirectUri!, // TODO: get this from client app
+    scopes: [],
+    user_agent: "uma-connect",
+    request_options: {
+      base_url: getNwcUmaDomain(uma),
+    },
+  });
+
+  let resultToken: UmaAuthToken;
+  if (token?.refreshToken) {
+    authClient.token = {
+      access_token: token.accessToken,
+      refresh_token: token.refreshToken,
+      expires_at: token.expiresAt,
+    };
+
+    resultToken = (await authClient.refreshAccessToken()).token as UmaAuthToken;
+  } else {
+    authClient.code_verifier = codeVerifier!; // TODO: where do we get this?
+    const code = new URLSearchParams(redirectUri!).get("code") || "";
+
+    resultToken = (await authClient.requestAccessToken(code)).token as UmaAuthToken;
+  }
+
+  return {
+    codeVerifier: "",
+    token: {
+      accessToken: resultToken.access_token,
+      refreshToken: resultToken.refresh_token,
```

Comment:
We probably need to store both the refresh token expiry (which is when the whole nwc connection expires) and the access token expiry. See [this section](https://docs.google.com/document/d/1__GDaJm0N6Iw14COeEu-FoJy8E-vphIkA6yRJN50k84/edit#heading=h.b4cpqw3lu472) of the design doc.

Reply:
hm i don't see two separate expires_at props for refresh/access. I do see in your PR you're expecting this nwc_expires_at prop through https://github.com/jklein24/bitcoin-connect/blob/feat/crosscurrency/src/connectors/UMANWCConnector.ts#L110C7-L110C21

Reply:
Yeah, see the `Token & {` on the first line there. The Token type has its own expiry which is for the access token.

Reply:
the token object includes both the access and refresh tokens; are you saying the expires_at prop is different depending on if refreshAccessToken or requestAccessToken is called?

Reply:
Not exactly. You get both back from the server. `nwc_expires_at` is the refresh token expiration and `expires_in` is the access token expiration: https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/. `nwc_expires_at` is our own thing of course.